### DISCUSSION TITLE
default templates loading accompanying style files

### DISCUSSION THREAD

**Discussion by iwelch** [jgm/pandoc#9974] [Q&A]
I am staring at the `default.latex` and `preamble.html` files.

Ideally, I would like to have any generated file for `xyz.md` in a pandoc conversion to html include the `xyz.css` and `xyz.js` files IF THEY EXIST. 

Similarly, I would like to have the generated file `xyz.md` in a pandoc conversion to latex include the `xyz.sty` file IF IT EXISTS.

this could be accomplished by slurping or referencing.  alas, to do even referencing, I need to have a variable that holds the basename of the file that is being processed, here `xyz` .

easy?

**Comment by bpj** [comment]
Seems like a job for a Makefile:

``````make
xyz.html: xyz.md
ifneq (,$(wildcard xyz.css))
	COND_OPTS := --css=xyz.css
endif
ifneq (,$(wildcard xyz.yaml))
	COND_OPTS := $(COND_OPTS) -dxyz.yaml
endif
	pandoc $(COND_OPTS) $< -so $@
``````

or with a bit of metaness:

``````make
ifeq ($(BASE),)
	$(error No BASE variable specified)
endif

$(BASE).html: $(BASE).md
ifneq (,$(wildcard $(BASE).css))
	COND_OPTS := --css=$(BASE).css
endif
ifneq (,$(wildcard $(BASE).yaml))
	COND_OPTS := $(COND_OPTS) -d$(BASE).yaml
endif
	pandoc $(COND_OPTS) $< -so $@

$(BASE).ltx: $(BASE).md
ifneq (,$(wildcard $(BASE).sty))
	COND_OPTS := -V header-includes='\usepackage{$(BASE)}'
endif
	pandoc -w latex $(COND_OPTS) $< -so $@
``````

Invoke with `make BASE=xyz`

You get the idea. The LaTeX bit is untested but should work.

The usual caveats about leading spaces in displayed code vs. tabs in a real Makefile apply!

**Comment by iwelch** [comment]
thx, bpj.  this may work, but seems quite elegant.  it would be much nicer if pandoc itself were to populate a few variables that one could then use in a default templates.

```
\IfFileExists{$basename.sty}{\usepackage{$basename}}{}
```

I guess what I really have then is a suggestion for a new feature.  if J does not read this forum, maybe I need to post it in the other one.

**Comment by badumont** [comment]
If you define "basename" in your metadata header/file, then you can use in your template:

     \IfFileExists{$basename$.sty}{\usepackage{$basename$}}{}

(Variables are written between two dollars in template files.)

**Comment by tarleb** [comment]
Keep in mind that pandoc allows multiple input files, in which case `basename` would be a list.

I tend to agree with @bpj  in that I believe that this is out of scope for pandoc and should be solved via alternative means.

**Reply by tarleb** [reply]
To offer another approach: this could be done in a custom writer. The list of input files is available through `PANDOC_STATE.input_files`. That'd probably be my tool of choice for this specific issue, because then you could also modify the template programmatically.

**Reply by bpj** [reply]
(I commented on this but for some reason it didn't show up. Apologies if this makes a duplicate!)

It occurred to me that being able to interpolate environment variables anywhere in defaults files might be a useful approach for cases like this.

**Reply by iwelch** [reply]
and the implementation in the core of pandoc would be very cheap --- just setting one global variable at the moment that the file is read in.  a "one-liner".

**Comment by iwelch** [comment]
thank you, everyone.  I guess a preprocessor that inserts into yaml the file could do the job relatively elegantly.  I will do this.  still, I would think that adding $basename$ automatically per the currently processed filename this would not necessary a bad feature to add.

**Reply by bpj** [reply]
Building a command line is generally safer and easier than modifying a YAML file. I certainly wouldn't do it with simple regex substitution on the whole file; YAML syntax is too finicky to do that safely. You basically have to parse the YAML, walk it (a library could probably help with that) and modify individual values, then serialize it again. I used to use a Perl script — actually configurable through a YAML file! — which built a command line. Defaults files mostly removed the need for it, but such an approach could be useful for a case like this.

**Reply by jgm** [reply]
Another approach could be adding a doctemplates "pipe" that takes the basename given a full filename.  This could then be used in the templates.
https://pandoc.org/MANUAL.html#pipes
This change would need to be made in jgm/doctemplates.

**Reply by iwelch** [reply]
indeed.  however, basename would seem such a nice feature and so easy to implement that I would throw it into the main feature set.  everytime a user has a `myfile.md`, the latex template would include myfile.sty, the html template would include myfile.css and myfile.js, etc., *if* these files exist.  that would be lovely for many users.  it would also probably reduce the need to tinker with template files.
